src/micro.c

```diff
@@ -0,0 +1,680 @@
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Description, cfg->Description));
+    MICRO_CALL(err, micro_clone_endpoint_config(&new_cfg->Endpoint, cfg->Endpoint));
+    if (err != NULL)
+    {
+        micro_free_cloned_service_config(new_cfg);
+        return err;
+    }
+
+    *out = new_cfg;
+    return NULL;
+}
+
+void micro_free_cloned_service_config(microServiceConfig *cfg)
+{
+    if (cfg == NULL)
+        return;
+
+    // the strings are declared const for the public, but in a clone these need
+    // to be freed.
+    NATS_FREE((char *)cfg->Name);
+    NATS_FREE((char *)cfg->Version);
+    NATS_FREE((char *)cfg->Description);
+    micro_free_cloned_endpoint_config(cfg->Endpoint);
+    NATS_FREE(cfg);
+}
+
+static void
+on_connection_closed(natsConnection *nc, void *closure)
+{
+    microService *m = (microService *)closure;
+    if (m == NULL)
+        return;
+
+    microService_Stop(m);
+
+    if (m->prev_on_connection_closed != NULL)
+    {
+        (*m->prev_on_connection_closed)(nc, m->prev_on_connection_closed_closure);
+    }
+}
+
+static void
+on_connection_disconnected(natsConnection *nc, void *closure)
+{
+    microService *m = (microService *)closure;
+
+    if (m == NULL)
+        return;
+
+    // <>/<> TODO: Should we stop the service? Not 100% how the Go client does
+    // it.
+    microService_Stop(m);
+
+    if (m->prev_on_connection_disconnected != NULL)
+    {
+        (*m->prev_on_connection_closed)(nc, m->prev_on_connection_disconnected_closure);
+    }
+}
+
+static void
+on_error(natsConnection *nc, natsSubscription *sub, natsStatus s, void *closure)
+{
+    microService *m = (microService *)closure;
+    microEndpoint *ep = NULL;
+    microError *err = NULL;
+    bool our_subject = false;
+
+    if ((m == NULL) || (sub == NULL))
+        return;
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; (!our_subject) && (ep != NULL); ep = ep->next)
+    {
+        if (micro_match_endpoint_subject(ep->subject, sub->subject))
+        {
+            break;
+        }
+    }
+    micro_unlock_service(m);
+
+    if (m->cfg->ErrHandler != NULL)
+    {
+        (*m->cfg->ErrHandler)(m, ep, s);
+    }
+
+    if (ep != NULL)
+    {
+        err = microError_Wrapf(micro_ErrorFromStatus(s), "NATS error on endpoint %s", ep->name);
+        micro_update_last_error(ep, err);
+        microError_Destroy(err);
+    }
+
+    // <>/<> TODO: Should we stop the service? The Go client does.
+    microService_Stop(m);
+
+    if (m->prev_on_error != NULL)
+    {
+        (*m->prev_on_error)(nc, sub, s, m->prev_on_error_closure);
+    }
+}
+
+static microError *
+wrap_connection_event_callbacks(microService *m)
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_getClosedCallback(&m->prev_on_connection_closed, &m->prev_on_connection_closed_closure, m->nc));
+    IFOK(s, natsConn_setClosedCallback(m->nc, on_connection_closed, m));
```

Comment:
Yes, so I don't seem to see that callbacks are chained more than one for `micro` and one for the user.